app/_main.py

```diff
@@ -35,13 +39,55 @@
         raise HTTPException(status_code=404, detail="Item not found")
 
 
-@app.get("/questions/{question_id}")
-async def read_question(question_id: int):
-    # keep a memo of questions served, so there are no repeats for a given session
-    # e.g., "questions_served.append({template: 4, game_id: 45})"
-    return {"question_id": question_id}
+@app.get("/questions/")
+async def read_question(asked: str = None):
+    "asked is a comma separated list of question ids that have been asked"
+    # memo passed between client & server ensures questions aren't repeated    
+    asked_memo = set(asked.split(",") if asked is not None else ())

```

Comment:
This is a really good point and it dovetails with something I've been thinking about already. Right now, the list of asked questions gets passed back and forth between client and server, and in the `client->server` case it's a url parameter. It seems like a bad idea to pass a url parameter that's like 1000 char long, which isn't too hard to imagine given that my question ids are often 6 or 7 chars. I could use base8 or something to shrink them a bit, bit it's still an issue. What I was thinking was capping the size of the list to say 50 games and then figuring forgetting older ones and allowing them to repeat (since remembering an answer 50+ questions later seems... hard), or even capping the length of the game to 50 or 100 questions...

Another option is to give the backend some persistent session state, in which case your solution could replace the list altogether, but seems non-RESTful.

I'll at least implement your suggestion, but if you have thoughts on the design decisions here I'd love to hear them

Reply:
If you decide not to implement sessions, you might consider:
- REST has major limitations that most shops will consider not worth pursuing for the sake of purity alone, so you might just do a POST on /questions with the state in the body or add a body to your GET, depending on what strikes your fancy
- What would it take to get RESTful sessions? You could maybe do something like POST to /transitions with `{"previousState": <previous state>, "previousAnswer": <answer>}` -> 200 `{"previousState": <same as request>, "previousAnswer": <same as request>, "nextState": <next state>, "nextQuestion": {…}}`.
- How much space does it really take to represent a complete game state? 6 bits can be represented by a single base64 digit. So if you've got 10 templates and 700 games, you could imagine an array with 7000 bit slots (0 means the question was not asked, 1 means it was). 7000/6 = *1167 base64 digits*. So, it seems borderline to me. If you can guarantee your templates won't grow very much, you might be able to get away with it.

Reply:
I like your first and second ideas. Do you have an opinion on one of those options, vs. using server side sessions?

If I'm breaking the rules of RESTfulness, should I use a different paradigm altogether? Do u have feelings on graphQL?

Reply:
I would go with the first solution (do a POST) – seems the simplest. GraphQL seems ok to me, but I worry about how much power it gives clients to formulate inefficient backend queries.

Reply:
i havent implemented the POST solution for state yet, but i did implement the never-infinite-loop fix in the new commit